gwpy/io/datafind.py

```diff
@@ -316,40 +121,55 @@ def on_tape(*files):
         otherwise `False`
     """
     for path in files:
+        stat = os.stat(path)
         try:
-            if os.stat(path).st_blocks == 0:
+            if stat.st_blocks == 0:
                 return True
         except AttributeError:  # windows doesn't have st_blocks
             return False
     return False
 
 
-def _choose_connection(**datafind_kw):
-    if os.getenv('LIGO_DATAFIND_SERVER') or datafind_kw.get('host'):
-        from gwdatafind import connect
-        return connect(**datafind_kw)
-    if os.getenv('VIRGODATA'):
-        return FflConnection()
-    raise RuntimeError("unknown datafind configuration, cannot discover data")
+def _gwdatafind_module(**datafind_kw):
+    """Return the appropriate GWDataFind-like API based on the environment
+
+    This allows switching to the hacky `gwpy.io.ffldatafind` replacement
+    module to enable a GWDataFind-like interface for direct FFL data
+    discovery at Virgo.
+    """
+    # GWDataFind
+    if (
+        os.getenv('GWDATAFIND_SERVER')
+        or os.getenv('LIGO_DATAFIND_SERVER')
+        or datafind_kw.get('host')
+    ):
+        return gwdatafind
+
+    # FFL
+    try:
+        ffldatafind._get_ffl_basedir()
+    except KeyError:  # failed to discover FFL directories
+        raise RuntimeError(
+            "unknown datafind configuration, cannot discover data",
+        )
+    return ffldatafind
 
 
-def with_connection(func):
-    """Decorate a function to open a new datafind connection if required
+def _select_gwdatafind_mod(func):
+    """Decorate a function to see the right ``gwdatafind`` API.
 
-    This method will inspect the ``connection`` keyword, and if `None`
-    (or missing), will use the ``host`` and ``port`` keywords to open
-    a new connection and pass it as ``connection=<new>`` to ``func``.
+    This exists only to allow on-the-fly replacing of the actual `gwdatafind`
+    with :mod:`gwpy.io.ffldatafind` if it looks like we are trying to find
+    data from FFL files.
     """
     @wraps(func)
     def wrapped(*args, **kwargs):
-        if kwargs.get('connection') is None:
-            kwargs['connection'] = _choose_connection(host=kwargs.get('host'),
-                                                      port=kwargs.get('port'))
-        try:
-            return func(*args, **kwargs)
-        except HTTPException:
-            kwargs['connection'] = reconnect(kwargs['connection'])
+        # replace the 'gwdatafind' module in the function namespace
+        # with the API we need for this call
+        with mock.patch.dict(func.__globals__):
```

Comment:
I'm having a hard time understanding what this line does and why it's here.  Is this `with X` (vs. the more common `with X as Y`) a context manager?  Or is it just creating a mock dictionary, copied from `func.__globals__`, which is then modified (based on the result of the call to `_gwdatafind_module`) and returned along with `func`, so that `func.__globals__["gwdatafind"]` is only modified in this way for this call/instance of the function, rather than changing the value of `func.__globals__["gwdatafind"]` for subsequent calls/instances, which I *think* is what would happen if line 169 were left out?  The evanescence of `mock` objects makes it hard for me to predict what is going to happen to the modified `__globals__["gwdatafind"]` value (cf. the code containing `with patch.dict('os.environ'` [here](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch.dict)).  Is there any reason *not* to delete line 169 and just use lines 170-171 to define `func.__globals__["gwdatafind"]` and leave it at that value for this (call/instance/whatever)?

Reply:
And yes, `with X` does the same thing as `with X as Y` just without exposing any objects relating to the context manager (because we don't need `Y` here). It's the `with` equivalent of not capturing the output of a function into a variable.

Reply:
I worry a little that the replacement in `func.__globals__` in one call would cause trouble in later calls, for reasons unknown.  I also worry about the usage of `mock.patch.dict`, because I don't understand them well enough to know when their influence evaporates (as seen in the link I referenced).  And I also worry about the work that isn't getting done while I'm trying to understand this, so since it's your code and your PR, I'm willing to trust you and let it become your problem if it breaks or is already subtly broken.

Reply:
I think my (relevant) knowledge gap is knowing when the context exits.  At least I know a lot more about it now, and I could *probably* figure it out if I took the time.